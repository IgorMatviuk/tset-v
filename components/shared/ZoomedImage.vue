<template>
  <div ref="zoomWrapper" class="zoomed"></div>
</template>

<script>
import ImageZoom from 'js-image-zoom'

export default {
  props: {
    imageUrl: {
      type: String,
      required: true
    }
  },
  data() {
    return {
      hoverZoom: null
    }
  },
  mounted() {
    /* istanbul ignore else */
    if (this.$refs.zoomWrapper) {
      this.setupHoverZoom()
    }
  },
  beforeDestroy() {
    this.hoverZoom.kill()
    this.hoverZoom = null
  },
  methods: {
    setupHoverZoom() {
      const element = this.$refs.zoomWrapper
      const mqlXl = window.matchMedia('(min-width: 1200px)')

      /* istanbul ignore else */
      if (element && mqlXl.matches) {
        this.hoverZoom = new ImageZoom(element, {
          width: element.offsetWidth,
          height: element.offsetHeight,
          zoomWidth: element.offsetWidth,
          img: this.imageUrl,
          zoomPosition: 'original'
        })
      }
    }
  }
}
</script>

<style lang="scss">
.zoomed {
  position: absolute !important;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 1;
  width: 100%;
  height: 100%;
  overflow: hidden;
  cursor: pointer;

  img {
    pointer-events: none;
    opacity: 0;
  }
}
</style>
