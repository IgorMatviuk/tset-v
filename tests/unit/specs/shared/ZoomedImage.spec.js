import { shallowMount } from '@vue/test-utils'
import ImageZoom from 'js-image-zoom'

import ZoomedImage from '@/components/shared/ZoomedImage'

jest.mock('js-image-zoom', () => {
  return jest.fn().mockImplementation(() => {
    return {
      kill: jest.fn()
    }
  })
})

describe('ZoomedImage', () => {
  beforeEach(() => {
    ImageZoom.mockClear()
  })

  function createMatchMediaMock(matches) {
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches,
        media: query,
        addListener: jest.fn().mockImplementation((callback) => callback()),
        removeListener: jest.fn().mockImplementation((callback) => callback()),
        addEventListener: jest.fn().mockImplementation((callback) => callback())
      }))
    })
  }

  function createZoomedImage() {
    const wrapper = shallowMount(ZoomedImage, {
      props: {
        imageUrl: 'https://some-url.com'
      }
    })

    return wrapper
  }

  test('render zoomed image', () => {
    createMatchMediaMock(true)

    const wrapper = createZoomedImage()

    expect(wrapper.element).toMatchSnapshot()
    expect(ImageZoom).toBeCalledTimes(1)

    wrapper.destroy()
  })
})
