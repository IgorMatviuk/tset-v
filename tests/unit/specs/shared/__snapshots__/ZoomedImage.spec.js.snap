// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZoomedImage render zoomed image 1`] = `
<div
  class="zoomed"
/>
`;
